File: delete-me-when-done/aks-based-demo/python-based/components/ingress_ctl.py
from pulumi import ComponentResource, ResourceOptions, Output
import pulumi_kubernetes as k8s

class IngressCtlArgs:

    def __init__(self,
                 namespace="ingress-ctl",
                 ):

        self.namespace = namespace

class IngressCtl(ComponentResource):

    def __init__(self,
                 name: str,
                 args: IngressCtlArgs,
                 opts: ResourceOptions = None):

        super().__init__('custom:kubernetes:IngressCtl', name, {}, opts)
        opts.parent = self

        namespace = args.namespace

        # Set up ingress controller
        ingress_namespace = k8s.core.v1.Namespace(f"{name}-{namespace}",
            metadata=k8s.meta.v1.ObjectMetaArgs(
                name=namespace
            ),
            opts=opts)

        nginx_ingress_name = "ingress-ctl"
        nginx_ingress = k8s.helm.v3.Chart(nginx_ingress_name, k8s.helm.v3.ChartOpts(
            fetch_opts={
                'repo': "https://kubernetes.github.io/ingress-nginx"
            },
            chart="ingress-nginx",
            namespace=ingress_namespace.metadata.name,
            values= {
                'controller': {
                    'replicaCount': 1,
                    'ingressClassResource': {
                        'default': True
                    },
                    'nodeSelector': {
                        "beta.kubernetes.io/os": "linux"
                    },
                    'admissionWebhooks': {
                        'patch': {
                            'nodeSelector': {
                                "beta.kubernetes.io/os": "linux"
                            }
                        }
                    }
                },
                'defaultBackend': {
                    'nodeSelector': {
                        "beta.kubernetes.io/os": "linux"
                    }
                }
            }),
            opts)

        ingress_service = nginx_ingress.get_resource("v1/Service", f"{namespace}/{nginx_ingress_name}-ingress-nginx-controller")
        self.ingress_service_ip = ingress_service.status.load_balancer.ingress[0].ip

File: delete-me-when-done/aks-based-demo/components/ingressCtl.ts
import * as pulumi from "@pulumi/pulumi";
import * as k8s from "@pulumi/kubernetes";

export interface IngressCtlArgs {
    kubeconfig: pulumi.Input<string>;
}

export class IngressCtl extends pulumi.ComponentResource {
  public readonly ingressCtlIp: pulumi.Output<string>;

  constructor(name: string, args: IngressCtlArgs, opts?: pulumi.ComponentResourceOptions) {

    super("custom:resource:IngressCtl", name, args, opts);

    // Instantiate the k8s provider
    const k8sProvider = new k8s.Provider("k8sProvider", {
        kubeconfig: args.kubeconfig
    })

    // Set up ingress controller
    const namespace = "ingress-ctl";

    const ingressNamespace = new k8s.core.v1.Namespace(`${name}-${namespace}`, {
       metadata: {
        name: namespace,
       } 
    }, {provider: k8sProvider, parent: this});

    const ingressName = "ingress-ctl"
    const nginxIngress = new k8s.helm.v3.Chart(ingressName, {
        fetchOpts: {
            repo: "https://kubernetes.github.io/ingress-nginx" 
        },
        chart: "ingress-nginx",
        namespace: ingressNamespace.metadata.name,
        values: {
            'controller': {
                'replicaCount': 1,
                'ingressClassResource': {
                    'default': true
                },
                'nodeSelector': {
                    "beta.kubernetes.io/os": "linux"
                },
                'admissionWebhooks': {
                    'patch': {
                        'nodeSelector': {
                            "beta.kubernetes.io/os": "linux"
                        }
                    }
                }
            },
            'defaultBackend': {
                'nodeSelector': {
                    "beta.kubernetes.io/os": "linux"
                }
            }
        }, 

    }, {provider: k8sProvider, parent: this});
            

    const ingressService = nginxIngress.getResource("v1/Service", `${namespace}/${ingressName}-ingress-nginx-controller`);
    this.ingressCtlIp = ingressService.status.loadBalancer.ingress[0].ip
  }
}
